Enhance SemObserved

alyst opened the pull request:
Another set of changes from #193 related to *SemObserved*:

* each *SemObserved* implementation now stores *observed_vars* and provides `observed_vars(observed)` method.
  That makes things simpler internally and, I think, also for the SEM users -- in particular it is now easy to check the *SemSpecification*, *SemImply*, and *SemObserved* compatibility.
* the rules for data import are relaxed and streamlined (to simplify things for the user):
   * `obs_colnames` kwarg renamed to `observed_vars` for consistency with the other functions (but I would like to hear feedback on the better name).
      The idea is that now it should be more clear what this argument is for.
   * as before, `SemObserved` ctors accept both `specification` and `observed_vars`. But with this PR having `SEM.observed_vars(spec) != observed_vars` would not raise an error.
     Instead, `observed_vars` kwarg takes precedence in mapping the `data` matrix columns to the observed variable names, whereas the `spec` takes precedence in defining the order of the variables in the imported data
     (also, observed variables not referenced in `spec` will not be imported). So if the correct SEM specification is already defined, the user will not have to do an extra manual step to match the order of the columns
     in the data matrix.
   * automatically generated observed variable IDs have *"obs"* prefix (`:obs1`, `:obs2` etc instead of `:1`, `:2`), that could be tweaked with `observed_var_prefix` kw arg.
   * this logic is implemented by `prepare_data()` function that is used by all `SemObserved` ctors
* `SemObservedCovariation` is made a type alias of `SemObservedData{Nothing}` (i.e. `nothing` as *data* matrix) instead of being its own type.
   That reduces some code duplication.
* `SemObservedMissing` is refactored to incapsulate all properties of a missing pattern into the single `SemObservedMissingPattern` object, so `SemObservedMissing` just stores a vector of these objects.
   In the context of the missing pattern, the observed variables that have values are called *measured* (so there are `measured_cov`, `measured_mask` etc fields); the vars without values are termed *missing* (`miss_mask`).
   That refactoring simplifies some code for EM, FIML, and minLL; and is also required for further enhancements of EM & FIML.

Maximilian-Stefan-Ernst commented on src/frontend/fit/fitmeasures/minus2ll.jl:
```diff
@@ -31,74 +31,33 @@ minus2ll(minimum::Number, obs, imp::Union{RAM, RAMSymbolic}, loss_ml::SemWLS) =
-)
-    F = 0.0
+    # fit EM-based mean and cov if not yet fitted
+    # FIXME EM could be very computationally expensive
```
Is this FIXME still relevant / could you explain it?

alyst replied:
I actually added it here.
I recall, at some point you correctly noted that forcing the EM run at *SemObservedMissing* construction is not a good idea, since it is an expensive operation that the user potentially doesn't even need.
Here it is somewhat similar: we implicitly trigger EM calculations within some innocent-looking fit measure call (minus2ll is used by the other measures too), the user may not expect that this could be very expensive.

For large problems and noisy data EM can take minutes, if not hours, to complete.
I think EM should be triggered explicitly, at least for larger problems.
There are some EM computation-related updates in my staging branch -- reviewing them would be a good moment to make the decision.

Maximilian-Stefan-Ernst commented on src/observed/covariance.jl:
```diff
@@ -1,128 +1,50 @@
-
-## Additional keyword arguments:
-- `spec_colnames::Vector{Symbol} = nothing`: overwrites column names of the specification object
+Type alias for [`SemObservedData`](@ref) that has mean and covariance, but no actual data.
```
I think we should maybe still keep some of the documentation here, to explain which keyword arguments can / should be passed to the constructor defined below.

alyst replied:
I have returned the revised the docstring for the `SemObservedCovariance` ctor.

Maximilian-Stefan-Ernst commented on src/observed/data.jl:
```diff
@@ -27,112 +25,36 @@ For observed data without missings.
+struct SemObservedData{D <: Union{Nothing, AbstractMatrix}} <: SemObserved
+    data::D
+    observed_vars::Vector{Symbol}
+    obs_cov::Matrix{Float64}
+    obs_mean::Vector{Float64}
     nsamples::Int
```
I'm thinking maybe we should not enforce `Float64` in case someone want's to use `Float32` for example, so maybe keeping this as type parameters would be a bit more flexible?

alyst replied:
Great point, especially since SemObservedMissing in this PR allows that.

Maximilian-Stefan-Ernst commented on src/observed/missing_pattern.jl:
```diff
@@ -0,0 +1,45 @@
+# data associated with the observed variables that all share the same missingness pattern
```
Just thinking maybe we should put the files "missing.jl", "missing_pattern.jl" and "EM.jl" in a folder "missing", and also bundle "data.jl" and "covariance.jl" in a different folder since one is now a type alias of the other.

alyst replied:
I have split "missing_pattern.jl" and "missing.jl" to allow EM to run before the *SemObservedMissing* is created, but since that logic would be revisited, they could be merged.

I don't have a strong opinion on the folder structure, but I guess I favor how easy it is to locate the file rather more than how closely the file structure follows the types hierarchy.
From that point, for the project of the SEM.jl size 1-level hierarchy should be fine. If one folder grows too big, it could be subdivided

Maximilian-Stefan-Ernst commented on src/observed/missing.jl:
```diff
@@ -9,193 +9,84 @@ mutable struct EmMVNModel{A, b, B}
     fitted::B
 end
 
+# FIXME type unstable
```
Is this also for revisiting later?

alyst replied:
Yes, it is related to the logic of computing EM.
As `obs_cov`/`obs_mean` could be called within `evaluate!` (not just FIML -- I am also using ML in combination with `SemObservedMissing`),
their type instability may significantly slow down the whole SEM fitting.

Maximilian-Stefan-Ernst replied:
Ah yes makes sense, that could be quite problematic...

Maximilian-Stefan-Ernst commented on src/observed/abstract.jl:
```diff
@@ -8,3 +8,143 @@ Rows are samples, columns are observed variables.
+            data_ordered = data
+        end
+        # make sure data_mtx is a dense matrix (required for methods like mean_and_cov())
+        data_mtx =
+            data_ordered isa DenseMatrix ? data_ordered : convert(Matrix, data_ordered)
```
It seems like the only subtypes of `DenseMatrix` are `Base.Experimental.Const{T, 2} where T` and `Matrix`, so maybe we can just do this:
```suggestion
        data_mtx = convert(Matrix, data_ordered)
```

Maximilian-Stefan-Ernst commented on test/unit_tests/data_input_formats.jl:
```diff
@@ -72,17 +84,17 @@ function test_observed(
     end
 
     # FIXME actually, SemObserved should not use meanstructure and always provide obs_mean()
```
This changed now, right? SemObserved does not use `meanstructure` anymore?

alyst replied:
Yes, that's true. We can remove it now from the unit tests, but actually the next PR I plan should be addressing #229.
So it would be, sort of, natural to remove it then (in particular because now passing `meanstructure` to `SemObserved` has no effect, but allowed -- but after the change it will generate an error).

Maximilian-Stefan-Ernst commented on src/frontend/specification/Sem.jl:
```diff
@@ -3,6 +3,7 @@
 ############################################################################################
 
 function Sem(;
+    specification = RAMMatrices,
```
I think we should make `ParameterTable` the default here - I know you are relying on the RAM notation to specify your models, but I believe that most users of the package (~90%) would rather use the `ParameterTable` specification.

I'm also thinking about always storing the specification object in the `Sem` model, maybe as it's own field in `Sem` - what do you think about that?

alyst replied:
I can change it, but actually specifying the class is no-op, as internally most SEM terms convert it into *RAMMatrices* anyway.
As for a separate specification object -- I am just not sure what additional features it will enable in addition to the existing mechanisms of conversion *DataFrame* into *SemSpecification* and *SemFit* into *DataFrame*?

Maximilian-Stefan-Ernst replied:
I think it would be nice to change that to `ParameterTable` - I'll write something below on the specification object.

alyst replied:
Fixed (also for `SemFiniteDiff` below)

Maximilian-Stefan-Ernst replied:
Thanks, I also added a keyword-only constructor for `ParameterTable` - so this works:
```
model = Sem(
    observed_vars = observed_vars,
    latent_vars = latent_vars,
    graph = graph,
    data = data,
    loss = SemML,
    meanstructure = false,
)
```

Maximilian-Stefan-Ernst commented on src/observed/abstract.jl:
```diff
@@ -8,3 +8,142 @@ Rows are samples, columns are observed variables.
+    data::Union{AbstractDataFrame, AbstractMatrix, Nothing},
+    observed_vars::Union{AbstractVector, Nothing},
+    spec::Union{SemSpecification, Nothing},
+    nobserved_vars::Union{Integer, Nothing} = nothing;
```
About the argument `nobserved_vars` - do we need this? I think we could maybe drop it, because in `SemObservedCovariance` it is called like this:

```
    if !isnothing(observed_vars)
        length(observed_vars) == nvars || throw(
            DimensionMismatch(
                "The length of the observed_vars $(length(observed_vars)) does not match the size of the covariance matrix $(size(obs_cov))",
            ),
        )
    end

    _, obs_vars, obs_vars_perm =
        prepare_data(nothing, observed_vars, specification, nvars; observed_var_prefix)
```

i.e. the check is done already before calling the function, and in `SemObservedData` it is not even passed:

```
function SemObservedData(;
    data,
    observed_vars::Union{AbstractVector, Nothing} = nothing,
    specification::Union{SemSpecification, Nothing} = nothing,
    observed_var_prefix::Union{Symbol, AbstractString} = :obs,
    kwargs...,
)
    data, obs_vars, _ =
        prepare_data(data, observed_vars, specification; observed_var_prefix)
    obs_mean, obs_cov = mean_and_cov(data, 1)

    return SemObservedData(data, obs_vars, obs_cov, vec(obs_mean), size(data, 1))
end
```

I'm thinking about the possibility to create a `Sem` without any `data` attached to it, to update the `data` later - this would be helpful for simulation studies, because at the moment, you always have to pass some mock data to a `Sem`. But because of this check in `prepare_data`, something like this errors:

```
model = Sem(
    specification = par_table,
    data = nothing,
    observed_vars = observed_vars,
    nobserved_vars = 11,
    loss = SemML,
    meanstructure = false,
)
```

`Cannot infer observed variables from provided inputs.`

Maximilian-Stefan-Ernst replied:
I opened an idea of how to implement this as a PR to this branch (didn't just want to commit it bcs it's not really thought out, just so you can see what you think about it)

alyst replied:
* Actually, there was a bug, thanks for catching it: when *data = nothing* and *observed_vars* was provided, it should not have raised an "Cannot infer observed variables" exception.
* I agree that `nobserved_vars` is somewhat redundant -- it is only required for the corner case when no data, observed_vars and no specification are provided.
  So now I changed `prepare_data()` -- the `nobserved_vars` is removed, but it now accepts the size of the data matrix as the `data` argument.
  Maybe it makes the logic a bit cleaner.
* At the moment `SemObservedData()` does not accept `data = nothing`, but this could be changed.
  Or you can implement `SemSimulatedData` that explicitly does not provide `obs_cov()` and `obs_mean()` at all (throws an error).

Maximilian-Stefan-Ernst commented:
Thanks again for this PR, I really like the streamlining of FIML and how the missing patterns are now computed/stored/used.

I also like the changes to observed - I'm still thinking about whether we should have a field `specification` in `Sem` - ATM, the `default` workflow includes having a separate specification object (mostly a `ParameterTable`), then creating a model from it, and then updating the partable with fitted estimates - it seems like this could be streamlined.

Maximilian-Stefan-Ernst commented on src/observed/abstract.jl:
```diff
@@ -8,3 +8,142 @@ Rows are samples, columns are observed variables.
+
+    # subset the data, check that obs_vars matches data or guess the obs_vars
+    if data isa AbstractDataFrame
+        if !isnothing(obs_vars_reordered) # subset/reorder columns
```
Should we throw an error if someone does something like

```
model = Sem(
    specification = par_table,
    data = example_data("political_democracy"),
    observed_vars = observed_vars[randperm(11)],
    loss = SemML,
    meanstructure = false,
)
```

i.e. pass a `DataFrame` and `observed_vars` with a different order of variables - bcs at the moment, `observed_vars` is just ignored in this case. I'm not sure there are any cases where it makes sense to pass both a `DataFrame` and `observed_vars`, since the data frame has names that should be used to reorder according to the `specification` object.

alyst replied:
I have added a commit that shows a warning in such cases (DataFrame, mismatch in `observed_vars` and `observed_vars(specification)`).
I think we write in the docstrings that the specification has a precedence in defining the final order of the variables, so it is not necessary to throw an error.
If the user thinks that *observed_vars* kw could be used to rename the *DataFrame* columns, that warning should clarify the confusion.

Maximilian-Stefan-Ernst replied:
Great, thanks.

alyst commented:
Thank you for the review -- thorough and helpful as usual!

> I'm still thinking about whether we should have a field `specification` in `Sem` - ATM, the "default" workflow includes having a separate specification object (mostly a `ParameterTable`), then creating a model from it, and then updating the partable with fitted estimates - it seems like this could be streamlined.

To me the efficiency of the model inference and the cleanliness of the user API are the priorities in the design of `Sem` types.
In that context *ParameterTable* is not a primary candidate for *SemSpecification* implementation.
I also do export the results of the SEM inference, but *ParameterTable* is just a transient intermediate before the conversion of *SemFit* into a *DataFrame*,
because *ParameterTable* lacks data manipulation capabilities of the *DataFrame*, and it does not support model-specific columns (in my case *observed_vars* are experiments, which have different metadata associated with them).
Theoretically, the conversion to/from *DataFrame* or the generic *Tables.jl* object into *SemSpecification* could be further streamlined.

Also, I am mostly working with SEM ensemble/multigroup models. Does your suggestion imply having specification objects both at the *SemEnsemble* level and *Sem* levels?
(Actually, the next big PR I plan merges *SemEnsemble* into *Sem*, which would then contain multiple SEM loss terms as well as multiple regularization terms -- so that problem goes away)

Maximilian-Stefan-Ernst commented on src/observed/abstract.jl:
```diff
@@ -8,3 +8,146 @@ Rows are samples, columns are observed variables.
+        if !isnothing(obs_vars_reordered) # subset/reorder columns
+            data = data[:, obs_vars_reordered]
+            if obs_vars_reordered != obs_vars
+                @warn "The order of variables in observed_vars argument does not match the order of observed_vars(specification). The specification order is used"
```
```suggestion
                @warn "The order of variables in observed_vars argument does not match the order of observed_vars(specification). The specification order is used."

```

Maximilian-Stefan-Ernst commented:
My thoughts about having this specification field were that it would just simplify things for the user bcs they don't have to manage a separate specification object. I expect that the overwhelming majority would have a partable to convert to a DataFrame, and they always have to create a partable -> create the model, and then at the end, update partable -> convert to data frame. Assuming the partable is living in the model, we could have a function `export_parameters` that just gives the DataFrame, and could make all the updating of the partable internally.

But this is not high priority right now, so after the last open comment is resolved, I would merge this 🥳

alyst commented:
I've changed the *Sem* default specification type to *ParameterTable*. I think that was the last remaining item from the review, so I have cleaned up the commit history a bit.

> My thoughts about having this specification field were that it would just simplify things for the user bcs they don't have to manage a separate specification object. 

I think we actually agree on how the workflow should look from the user perspective: most of the time the user works with the parameter table representation - both for specifying the model and processing the results.
And, most likely, for the user it is most convenient to work with the parameter table as a *DataFrame* object, since it has all data manipulation tools that *ParameterTable* type does not have.
My point was just that since the conversion to/from *ParameterTable* representation is trivial and cheap, we don't actually have to store it, we can just generate it on the fly.
For the user it is fairly easy to integrate that minimal *ParameterTable* with just the parameter estimates into the existing SEM data frame by using `innerjoin()` -- it also automatically takes care of the integrity.
With that we also avoid the information flow loop: we don't store the results of model fitting in the same object that was used as an input.

Maximilian-Stefan-Ernst commented:
Thanks! I added one last commit, if this one is fine with you, I would merge.

Also, the workflow you suggest seems like a good idea - so maybe writing a function like `export_params(sem_fit)` that generates a `DataFrame` on the fly, and have functions like `sem_summary(sem_fit)` create the partable on the fly.

alyst commented:
> Thanks! I added one last commit, if this one is fine with you, I would merge.

Thank you!

alyst commented:
I've realized that the tests were failing with the commit that switches Sem to ParameterTable by default. I've fixed it, but I think it is another indication that SEM.jl should move away from kw-arg based constructors.
Julia does not dispatch based on the names of kw used or on their types, so if there is a kw-only ctor for a type, it has to be the single kw-only ctor.
But that puts limitations on the how extendable is the package.

alyst commented:
*ProximalAlgorithms.jl* were update to 0.7 yesterday (from 0.5.4).
It looks like the new version has changed how user-defined functions should provide their gradients, so I had restricted ProxAlgorithms to 0.5 for now.

Maximilian-Stefan-Ernst commented:
Thanks a lot for the fix!
